src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Utilities/NullableHelpers.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis;
+
+internal static class NullableHelpers
+{
+    /// <summary>
+    /// Gets the declared symbol and root operation from the passed in declarationSyntax and calls <see
+    /// cref="IsSymbolAssignedPossiblyNullValue"/>. Note that this is bool and not bool? because we know that the symbol
+    /// is at the very least declared, so there's no need to return a null value. 
+    /// </summary>
+    public static bool IsDeclaredSymbolAssignedPossiblyNullValue(
+        ISemanticFacts semanticFacts,
+        SemanticModel semanticModel,
+        SyntaxNode declarationSyntax,
+        CancellationToken cancellationToken)
+    {
+        var declaredSymbol = semanticModel.GetRequiredDeclaredSymbol(declarationSyntax, cancellationToken);
+        var declaredOperation = semanticModel.GetRequiredOperation(declarationSyntax, cancellationToken);
+
+        var rootOperation = declaredOperation;
+
+        // Walk up the tree to find a root for the operation
+        // that contains the declaration
+        while (rootOperation is not IBlockOperation &&
+            rootOperation.Parent is not null)
+        {
+            rootOperation = rootOperation.Parent;
+        }
+
+        return IsSymbolAssignedPossiblyNullValue(
+            semanticFacts, semanticModel, rootOperation, declaredSymbol, rootOperation.Syntax.Span, includeDeclaration: true, cancellationToken) == true;
+    }
+
+    /// <summary>
+    /// Given an operation, goes through all descendent operations and returns true if the symbol passed in
+    /// is ever assigned a possibly null value as determined by nullable flow state. Returns
+    /// null if no references are found, letting the caller determine what to do with that information
+    /// </summary>
+    public static bool? IsSymbolAssignedPossiblyNullValue(
+        ISemanticFacts semanticFacts,
+        SemanticModel semanticModel,
+        IOperation rootOperation,
+        ISymbol symbol,
+        TextSpan span,
+        bool includeDeclaration,
+        CancellationToken cancellationToken)
+    {
+        var hasReference = false;
+
+        using var _ = ArrayBuilder<IOperation>.GetInstance(out var stack);
+        stack.Push(rootOperation);
+
+        while (stack.TryPop(out var operation))
+        {
+            if (!span.IntersectsWith(operation.Syntax.Span))
```

Comment:
>if (!span.IntersectsWith(operation.Syntax.Span)) [](http://example.com/codeflow?start=12&length=48)

I usually like this better when the check is used to avoid putting into the stack (ie, if this were inside the for loop on line 129)

Reply:
oh, but maybe that doesn't work in this case as you have other locations that are pushing onto the stack (like line 90)

Reply:
yes.